src/linear_algebra/matrix/pos_def.lean

```diff
@@ -29,6 +27,42 @@ open_locale matrix
 def pos_def (M : matrix n n R) :=
 M.is_hermitian ∧ ∀ x : n → R, x ≠ 0 → 0 < dot_product (star x) (M.mul_vec x)
 
+/-- A matrix `M : matrix n n R` is positive semidefinite if it is hermitian
+   and `xᴴMx` is nonnegative for all `x`. -/
+def pos_semidef (M : matrix n n R) :=
+M.is_hermitian ∧ ∀ x : n → R, 0 ≤ dot_product (star x) (M.mul_vec x)
+
+lemma pos_def.pos_semidef {M : matrix n n R} (hM : M.pos_def) : M.pos_semidef :=
+begin
+  refine ⟨hM.1, _⟩,
+  intros x,
+  by_cases hx : x = 0,
+  { simp only [hx, zero_dot_product, star_zero] },
+  { exact le_of_lt (hM.2 x hx) }
+end
+
+@[simp] lemma pos_semidef.minor {M : matrix n n R} (hM : M.pos_semidef) (e : m ≃ n):
```

Comment:
It's true for injective functions, but harder to prove. For noninjective functions I suspect it does not hold.

Reply:
Ok, for noninjective functions definitely not:
[[8, 2], [2, 0.5]] is pos. semidef., but [[8, 2, 2], [2, 0.5, 0.5], [2, 0.5, 0.5]] is not.

Reply:
I would guess the proof with injective function is a trivial modification of the current proof; you just obtain `e.symm` through `hf.has_left_inverse` where `hf : function.injective f`

Reply:
In the current proof you can see that I need both `equiv.symm_apply_apply` and `equiv.apply_symm_apply`. So I am relying on the `e.symm` to be a right inverse as well. Another annoyance with `hf.has_left_inverse` is that it needs `m` to nonempty.

Essentially, what I would need to do is to define a vector `y : n → R` that maps `i : n` to `f⁻¹(i)` if it exists and to `0` otherwise.